From 856236ad8a0c6fe507b7399c230706b0ee4a206f Mon Sep 17 00:00:00 2001
From: Pontus Fuchs <pontus.fuchs@gmail.com>
Date: Fri, 28 Sep 2012 20:33:15 +0200
Subject: [PATCH 57/68] Don't decrement tx pending counter if HW is not up

On rmmod followed by and insmod without USB unplug the
WDCMSG_SEND_COMPLETE event might show up after the insmod.
This will lead to negative tx pending counter.
---
 ar5523.c |    5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

--- a/drivers/net/wireless/ar5523.c
+++ b/drivers/net/wireless/ar5523.c
@@ -276,7 +276,10 @@ static void ar5523_cmd_rx_cb(struct urb
 	case WDCMSG_SEND_COMPLETE:
 		ar5523_dbg(ar, "WDCMSG_SEND_COMPLETE: %d pending\n", 
 			atomic_read(&ar->tx_nr_pending));
-		ar5523_data_tx_pkt_put(ar);
+		if (!test_bit(AR5523_HW_UP, &ar->flags))
+			ar5523_dbg(ar, "Unexpected WDCMSG_SEND_COMPLETE\n");
+		else
+			ar5523_data_tx_pkt_put(ar);
 		break;
 
 	case WDCMSG_TARGET_START:
